lib/einhorn/command.rb

```diff
@@ -379,6 +385,54 @@ def self.prepare_child_process
       Einhorn.renice_self
     end
 
+    def self.setup_parent_watch(expected_ppid, options={})
+      if Einhorn::State.kill_children_on_exit then
+        if options.fetch(:will_exec, false) then
+          # This allows einhorn workers to properly handle parent death that
+          # occurs after the parent check below, but before the exec syscall.
+          #
+          # The signal handler will be reset to the default (ignore) on exec.
+          #
+          # There is still a one-ruby-instruction race condition window during
+          # the handling of exec, see below.
+          Signal.trap("USR2") do
+            Einhorn.log_error("Parent process died before exec")
+            exit(1)
+          end
+        end
+
+        begin
+          Einhorn::Prctl.set_pdeathsig("USR2")
+          if Process.ppid != expected_ppid then
+            Einhorn.log_error("Parent process died before we set pdeathsig; cowardly refusing to exec child process.")
+            exit(1)
+          end
+
+          # NB: There is an unfixable race condition at the exec. If our parent
+          # dies and we receive the USR2 signal, then ruby will enqueue the
+          # signal to be handled after the VM instruction that is currently
+          # running. If we get unlucky and this VM instruction happens to be the
+          # one that does the exec syscall, we've permanently lost the fact that
+          # we should have died.
+          #
+          # This is unfixable given that the graceful shutdown
+          # signal is USR2; if it were a signal whose default behavior was to
+          # kill the process, like HUP, then we could tell the kernel to reset
+          # our handler for HUP to the default (meaning it would not go to ruby
+          # at all, the kernel would handle it as a SIGKILL), then it
+          # wouldn't matter whether we were in the exec ruby VM instruction or
+          # not, since ruby would not be responsible for handling the signal.
+          #
+          # Setting the handler for USR2 to kill the process would fix this, but
+          # would also leak an unexpected signal handler into the worker process
+          # in the normal case, which is unacceptable.
```

Comment:
wait, why can't we just set the `USR2` handler to the default (just kill the process)? the exec-ed worker is responsible for setting their own graceful shutdown `USR2` anyway, right?

also, the wrapping a paragraph up is a little weird